File: MCRRTImp/src/test/java/infrastructure/Vector2BasedImp/Vector2.java
package infrastructure.Vector2BasedImp;

import lab.mars.RRTBase.Dimension;
import lab.mars.RRTBase.MathUtil;
import lab.mars.RRTBase.Vector;

import java.util.Objects;

import static lab.mars.RRTBase.MathUtil.D2R;
import static lab.mars.RRTBase.MathUtil.R2D;

public class Vector2 extends Vector<Vector2> {

    private Dimension x;

    private Dimension y;


    public double x() {
        return x.value;
    }

    public double y() {
        return y.value;
    }

    public Vector2() {
        this(0, 0);
    }

    public Vector2(double x, double y) {
        super(x, y);
        this.x = dimensions[0];
        this.y = dimensions[1];
    }

    public double distance(Vector2 o) {
        double dx = x.value - o.x.value;
        double dy = y.value - o.y.value;
        return Math.sqrt(dx * dx + dy * dy);
    }

    public double distance2(Vector2 o) {
        double dx = x.value - o.x.value;
        double dy = y.value - o.y.value;
        return dx * dx + dy * dy;
    }

    public Vector2 normalize() {
        double length = len();
        x.value = x.value / length;
        y.value = y.value / length;
        return this;
    }

    public Vector2 cpy() {
        return new Vector2(x.value, y.value);
    }

    public double len() {
        return Math.sqrt(x.value * x.value + y.value * y.value);
    }

    public double len2() {
        return x.value * x.value + y.value * y.value;
    }

    public Vector2 set(Vector2 o) {
        this.x.value = o.x.value;
        this.y.value = o.y.value;
        return this;
    }

    public Vector2 subtract(Vector2 v) {
        x.value = x.value - v.x.value;
        y.value = y.value - v.y.value;
        return this;
    }

    public Vector2 translate(Vector2 v) {
        x.value = x.value + v.x.value;
        y.value = y.value + v.y.value;
        return this;
    }

    public double dot(Vector2 v) {
        return x.value * v.x.value + y.value * v.y.value;
    }

    public Vector2 scale(double scalar) {
        this.x.value = x.value * scalar;
        this.y.value = y.value * scalar;
        return this;
    }

    public Vector2 scale(double x, double y) {
        this.x.value *= x;
        this.y.value *= y;
        return this;
    }

    public Vector2 scale(Vector2 v) {
        this.x.value = x.value * v.x.value;
        this.y.value = y.value * v.y.value;
        return this;
    }

    public double angle(Vector2 o) {
        double value = this.dot(o) / (len() * o.len());
        if (value < -1) {
            value = -1;
        } else {
            if (value > 1) {
                value = 1;
            }
        }
        return Math.acos(value) * R2D;
    }

    @Override
    public Vector2 reverse() {
        this.x.value = - this.x.value;
        this.y.value = - this.y.value;
        return this;
    }

    public Vector2 rotate(double alpha) {
        double cos = Math.cos(alpha * D2R);
        double sin = Math.sin(alpha * D2R);
        double x_n = x.value * cos - y.value * sin;
        double y_n = x.value * sin + y.value * cos;
        x.value = x_n;
        y.value = y_n;
        return this;
    }

    public Vector2 lerp(Vector2 target, double coefficient) {
        double invert = 1.0f - coefficient;
        this.x.value = (x.value * invert) + (target.x.value * coefficient);
        this.y.value = (y.value * invert) + (target.y.value * coefficient);
        return this;
    }

    public boolean epsilonEquals(Vector2 other, double epsilon) {
        if (other == null) return false;
        if (Math.abs(other.x.value - x.value) > epsilon) return false;
        if (Math.abs(other.y.value - y.value) > epsilon) return false;
        return true;
    }

    public Vector2 zero() {
        this.x.value = 0;
        this.y.value = 0;
        return this;
    }

    @Override
    public String toString() {
        return "Vector2{" +
                "x.value=" + x.value +
                ", y.value=" + y.value +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vector2 vector2 = (Vector2) o;
        return MathUtil.epsilonEquals(vector2.x.value, x.value) &&
                MathUtil.epsilonEquals(vector2.y.value, y.value);
    }

    @Override
    public int hashCode() {

        return Objects.hash(x.value, y.value);
    }

    public Vector2 set(double x, double y) {
        this.x.value = x;
        this.y.value = y;
        return this;
    }
}

File: MCRRTImp/src/test/kotlin/infrastructure/Vector2BasedImp/Vector2.kt
package infrastructure.Vector2BasedImp

import lab.mars.MCRRTImp.base.*

import java.util.Objects


class Vector2 @JvmOverloads constructor(x: Double = 0.0, y: Double = 0.0) : Vector<Vector2>(x, y) {

    private val x: Dimension = dimensions[0]

    private val y: Dimension = dimensions[1]


    fun x(): Double {
        return x.value
    }

    fun y(): Double {
        return y.value
    }

    override fun distance(o: Vector2): Double {
        val dx = x.value - o.x.value
        val dy = y.value - o.y.value
        return Math.sqrt(dx * dx + dy * dy)
    }

    override fun distance2(o: Vector2): Double {
        val dx = x.value - o.x.value
        val dy = y.value - o.y.value
        return dx * dx + dy * dy
    }

    override fun normalize(): Vector2 {
        val length = len()
        x.value = x.value / length
        y.value = y.value / length
        return this
    }

    override fun cpy(): Vector2 {
        return Vector2(x.value, y.value)
    }

    override fun len(): Double {
        return Math.sqrt(x.value * x.value + y.value * y.value)
    }

    override fun len2(): Double {
        return x.value * x.value + y.value * y.value
    }

    override fun set(o: Vector2): Vector2 {
        this.x.value = o.x.value
        this.y.value = o.y.value
        return this
    }

    fun subtract(v: Vector2): Vector2 {
        x.value = x.value - v.x.value
        y.value = y.value - v.y.value
        return this
    }

    override fun translate(v: Vector2): Vector2 {
        x.value = x.value + v.x.value
        y.value = y.value + v.y.value
        return this
    }

    override fun dot(v: Vector2): Double {
        return x.value * v.x.value + y.value * v.y.value
    }

    override fun scale(scalar: Double): Vector2 {
        this.x.value = x.value * scalar
        this.y.value = y.value * scalar
        return this
    }

    fun scale(x: Double, y: Double): Vector2 {
        this.x.value *= x
        this.y.value *= y
        return this
    }

    override fun scale(v: Vector2): Vector2 {
        this.x.value = x.value * v.x.value
        this.y.value = y.value * v.y.value
        return this
    }

    override fun angle(other: Vector2): Double {
        var value = this.dot(other) / (len() * other.len())
        if (value < -1) {
            value = -1.0
        } else {
            if (value > 1) {
                value = 1.0
            }
        }
        return Math.acos(value) * R2D
    }

    override fun reverse(): Vector2 {
        this.x.value = -this.x.value
        this.y.value = -this.y.value
        return this
    }

    override fun rotate(angle: Double): Vector2 {
        val cos = Math.cos(angle * D2R)
        val sin = Math.sin(angle * D2R)
        val xN = x.value * cos - y.value * sin
        val yN = x.value * sin + y.value * cos
        x.value = xN
        y.value = yN
        return this
    }

    fun lerp(target: Vector2, coefficient: Double): Vector2 {
        val invert = 1.0f - coefficient
        this.x.value = x.value * invert + target.x.value * coefficient
        this.y.value = y.value * invert + target.y.value * coefficient
        return this
    }

    fun epsilonEquals(other: Vector2?, epsilon: Double): Boolean {
        if (other == null) return false
        if (Math.abs(other.x.value - x.value) > epsilon) return false
        return Math.abs(other.y.value - y.value) <= epsilon
    }

    override fun zero(): Vector2 {
        this.x.value = 0.0
        this.y.value = 0.0
        return this
    }

    override fun toString(): String {
        return "Vector2{" +
                "x.value=" + x.value +
                ", y.value=" + y.value +
                '}'.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val vector2 = other as Vector2?
        return (vector2!!.x.value epsilonEquals x.value) && (vector2.y.value epsilonEquals y.value)
    }

    override fun hashCode(): Int {

        return Objects.hash(x.value, y.value)
    }

    operator fun set(x: Double, y: Double): Vector2 {
        this.x.value = x
        this.y.value = y
        return this
    }

    override fun epsilonEquals(other: Vector2): Boolean {
        return epsilonEquals(other, 0.0001)
    }
}
